### ISSUE TITLE
refactor redrock spectra -> MPI shared memory input

### ISSUE THREAD

**Issue by sbailey** [desihub/redrock#36]
Currently `rrdesi_mpi` loads all spectra into rank 0 memory and then distributes into shared memory  for use by other MPI ranks.  For large input spectra files, this exceeds the memory quota for rank 0 and causes a crash even if the total memory is below the available memory on the node.

Refactor the spectra input to read in chunks and distribute to shared memory without requiring rank 0 to have all spectra simultaneously in non-shared memory.

This may be possible with `astropy.io.fits` using `memmap=True`, otherwise ok to use `fitsio` if needed.

**Comment by tskisner** [comment]
I have some standalone testing code for this which uses astropy.io.fits with memmap.  As I integrate that into redrock, I just wanted to describe the changes I'm going to do so that you can voice opinions if you have them.  The current existing code has:

1.  desi-specific read_spectra() function that reads multiple spectra files into a single collection of targets in memory.

2.  generic MPISharedTargets object that takes already-read targets and copies them into shared memory.
    - this class has a method _fill_target_dictionary() which computes the buffer offsets
    - constructor calls _fill_target_dictionary with the existing targets to compute the offsets
    - constructor then copies the targets into shared memory one at a time.


We could do this with classes and inheritance, but you requested that the I/O be done with free functions.  The proposed new structure is then:

1.  new desi-specific desi_target_offsets() function that opens files and computes the equivalent offsets as _fill_target_dictionary.

2.  move _fill_target_dictionary out of the MPISharedTargets class and change into a generic "target_offsets()" function that computes the offsets on existing in-memory target lists.

3.  Change MPISharedTargets constructor to now accept a target_offsets dictionary and NOT a list of targets in memory.  An instance of MPISharedTargets will now be constructed with a fixed size, but no contents.

4.  Add method to MPISharedTargets to insert a list of targets.  This is needed to support the current mode of operation where we have the targets in normal memory and want to copy them to shared memory.

5.  new desi-specific "read_spectra_shared()" function that takes a list of spectra files and an existing MPISharedTargets object.  This function will then read through the HDUs one at a time and copy the data directly into the raw buffers of MPISharedTargets.  It will NOT use the "insert" method mentioned above, since we want to pass through the file in a linear fashion one HDU at a time.  So we never have the info for a single target in normal memory at the same time.

To summarize, the new code will support the existing use case where targets are all in normal memory (pseudo code here):
```
targets = read_spectra(files)
offsets = target_offsets(targets)
shared = MPISharedTargets(offsets)
shared.insert(targets)
```
And it will also support the new desi way of reading targets in buffers:
```
offsets = desi_target_offsets(files)
shared = MPISharedTargets(offsets)
read_spectra_shared(shared, files)
```
Any comments from @sbailey or @julienguy especially welcome.

**Comment by julienguy** [comment]
That's good for me

**Comment by sbailey** [comment]
OK, we can try that.  Thanks for including the final snippets of example code.  If that works robustly, great.  

What I care most about is that the underlying algorithm code can use the targets as if they were a vanilla list of objects without knowing or caring that there is fancy shared memory stuff going on under the hood.  I think that is the case with what you propose, i.e. the messiness is contained to the loading, but not beyond that.  If that's not the case or if we find it problematic to debug/maintain/understand, I'd suggest an alternative:

Maintain the shared memory buffers per Target instead of a megabuffer in MPISharedTargets that couples the Targets to their container.  This approach works fine with multiprocessing but there might be some MPI limitation with having N>>1 smaller shared memory buffers instead of one large one.  To fill these, loop over the HDUs in parallel and load one target at a time (i.e. still not requiring all targets to be in non-shared memory simultaneously).

If you can implement this with astropy.io.fits using sharedmem, great.  If that is problematic, ok to use fitsio instead.

**Comment by sbailey** [comment]
implemented in PR #67
